File: codes/python/435.py
from typing import List


class Solution:
    def eraseOverlapIntervals(self, intervals: List[List[int]]) -> int:
        if not intervals:
            return 0

        intervals.sort(key=lambda x: x[1])

        n = len(intervals)
        # 首个区间
        right = intervals[0][1]
        ans = 1

        # 对于接下来的区间，选择大于 left 且 right 最小的即可
        for i in range(1, n):
            if intervals[i][0] < right:
                continue
            right = intervals[i][1]
            ans += 1

        # 因为得到的是不重叠区间的数量，因此需要我们减掉。
        return n-ans


if __name__ == '__main__':
    sol = Solution()

File: codes/cpp/435.cpp
#include "global.hpp"

class Solution
{
public:
    int eraseOverlapIntervals(vector<vector<int>> &intervals)
    {
        if (intervals.empty())
            return 0;

        // 排序
        sort(intervals.begin(), intervals.end(), [](const auto &u, const auto &v)
             { return u[1] < v[1]; });

        // 状态转移
        int n = intervals.size();
        vector<int> dp(n, 1);
        for (int i = 1; i < n; i++)
        {
            for (int j = 0; j < i; j++)
            {
                if (intervals[j][1] <= intervals[i][0])
                {
                    dp[i] = max(dp[i], dp[j] + 1);
                }
            }
        }
        return n - *max_element(dp.begin(), dp.end());
    }
};

int main()
{
    Solution sol;
    // assert(sol.Func() == xx);

    return 0;
}
